src/sage/rings/function_field/drinfeld_modules/anderson_motive.py

```diff
@@ -0,0 +1,1226 @@
+          the names of the vector of the canonical basis; if ``None``,
+          elements will be represented as row vectors
+
+        - ``normalize`` -- a boolean (default: ``True``)
+
+        The action of `\tau` on the Anderson motive will be given by
+        the matrix ``tau * (T - z)**(-twist)`` where `T` is the variable
+        of the function ring and `z` is its image in the `A`-field.
+
+        TESTS::
+
+            sage: A.<T> = GF(5)[]
+            sage: K.<z> = GF(5^3)
+            sage: M = AndersonMotive(A, K)
+            sage: type(M)
+            <class 'sage.rings.function_field.drinfeld_modules.anderson_motive.AndersonMotive_general_with_category'>
+
+        ::
+
+            sage: M1 = M.carlitz_twist(-2)
+            sage: M2 = AndersonMotive(A, M1.matrix())
+            sage: M1 is M2
+            True
+        """
+        AK = category.base()
+        K = AK.base_ring()
+
+        # We normalize the inputs
+        twist = ZZ(twist)
+        tau = tau.change_ring(AK)
+        if normalize:
+            divisor = category.divisor()
+            exponent = Infinity
+            for entry in tau.list():
+                if not entry:
+                    continue
+                e = 0
+                while entry.degree() > 0 and e < exponent:
+                    entry, R = entry.quo_rem(divisor)
+                    if R:
+                        break
+                    e += 1
+                exponent = e
+                if exponent == 0:
+                    break
+            if exponent is not Infinity and exponent > 0:
+                denom = divisor ** exponent
+                tau = tau.parent()([entry // denom for entry in tau.list()])
+                twist -= exponent
+
+        names = normalize_names(names, tau.nrows())
+        denominator = Factorization([(category.divisor(), twist)])
+        ore = category._ore_polring
+
+        #if (isinstance(K, FractionField_1poly_field)
+        #    and category.constant_coefficient() == K.gen()):
+        #    from sage.rings.function_field.drinfeld_modules.anderson_motive_rational import AndersonMotive_rational
+        #    cls = AndersonMotive_rational
+        #else:
+        cls = AndersonMotive_general
+
+        return cls.__classcall__(cls, tau, ore, denominator, names, category)
+
+    def __init__(self, mat, ore, denominator, names, category) -> None:
+        r"""
+        Initialize this Anderson motive.
+        """
+        OreModule.__init__(self, mat, ore, denominator, names, category)
+        self._initialize_attributes()
+
+    def _initialize_attributes(self):
+        r"""
+        Set the main attributes to this Anderson motive.
+
+        .. NOTE::
+
+            Separating this method from `__init__` makes it easier
+            to call it in subclasses.
+        """
+        self._tau = self._pseudohom.matrix()
+        if self._denominator:
+            self._twist = self._denominator[0][1]
```

Comment:
I think some pattern matching this might be better (for reading and for checking soundness at the same time):

```suggestion
            [(self._twist, _)] = self._denominator
```